tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala

```diff
@@ -54,14 +57,53 @@ object CodeGen {
       // of available processesors. This prevents LLVM from optimizing
       // across IR module boundary unless LTO is turned on.
       def separate(): Seq[Path] =
-        partitionBy(assembly, procs)(_.name.top.mangle).par
+        assembly
+          .groupBy { defn =>
+            val packageName =
+              defn.name.top.id.split("\\.").dropRight(1).mkString(".")
+            packageName
+          }
+          .par
           .map {
-            case (id, defns) =>
-              val sorted = defns.sortBy(_.name.show)
-              Impl(config, env, sorted).gen(id.toString, workdir)
+            case (pack, defns) =>
+              if (incCompilationContext != null) {
+                incCompilationContext.collectFromCurr(pack, defns)
+              }
+              if (incCompilationContext == null ||
+                  incCompilationContext.isChanged(pack)) {
+                val sorted = defns.sortBy(_.name.show)
+                val packagePrefix = config.workdir resolve
+                  Paths.get(
+                    pack
+                      .split(s"\\.")
+                      .dropRight(1)
+                      .mkString(File.separatorChar.toString)
+                  )
+                if (!packagePrefix.toFile.exists()) {
+                  packagePrefix.toFile.mkdirs()
+                }
+                val packagePath = pack
+                  .split(s"\\.")
+                  .mkString(File.separatorChar.toString)
+                Impl(config, env, sorted).gen(packagePath, workdir)
+              } else {
+                val packagePrefix = config.workdir resolve Paths.get(
+                    pack
+                      .split(s"\\.")
+                      .dropRight(1)
+                      .mkString(File.separatorChar.toString)
+                  )
+                  .toFile
+                  .toPath
```

Comment:
We can use path resolve + normalize to achieve the same effect
```suggestion
                val packagePrefix = config.workdir 
                 .resolve(pack.replace(".", File.separator))
                 .resolve("..")
                 .normalize
```